### PR TITLE
[SPARK-14807] Create a compatibility module

### PR DESCRIPTION
**Pull request by yhuai**
## What changes were proposed in this pull request?

This PR creates a compatibility module in sql (called `hive-1-x-compatibility`), which will host HiveContext in Spark 2.0 (moving HiveContext to here will be done separately). This module is not included in assembly because only users who still want to access HiveContext need it.
## How was this patch tested?

I manually tested `sbt/sbt -Phive package` and `mvn -Phive package -DskipTests`.

### CODE REVIEW

File: compatibility/pom.xml [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+                        <compilerArgs combine.children="append">
+                            <!-- This option is needed to suppress warnings from sun.misc.Unsafe usage -->
+                            <arg>-XDignore.symbol.file</arg>
+                        </compilerArgs>
+                    </configuration>
+                </plugin>
+            </plugins>
+        </pluginManagement>
+    </build>
+</project>
```
Comment (rxin) [inline compatibility/pom.xml]: add an extra line

**Comment by rxin** [comment]
I'd name this something like spark-1.x-compatibility (is dot allowed in the name?)

File: project/SparkBuild.scala [hunk 2]
```diff
@@ -254,7 +254,7 @@ object SparkBuild extends PomBuild {
   val mimaProjects = allProjects.filterNot { x =>
     Seq(
       spark, hive, hiveThriftServer, catalyst, repl, networkCommon, networkShuffle, networkYarn,
-      unsafe, testTags, sketch, mllibLocal
+      unsafe, testTags, sketch, mllibLocal, compatibility
```
Comment (yhuai) [inline project/SparkBuild.scala]: Looks like we should move sketch and compatibility out from this list once we release 2.0, right?

File: compatibility/pom.xml [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+  -->
+
+<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+         xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
+    <modelVersion>4.0.0</modelVersion>
+    <parent>
+        <groupId>org.apache.spark</groupId>
+        <artifactId>spark-parent_2.11</artifactId>
+        <version>2.0.0-SNAPSHOT</version>
+        <relativePath>../../pom.xml</relativePath>
```
Comment (JoshRosen) [inline compatibility/pom.xml]: I think this relative path is wrong. Probably should only have one `../`.

**Comment by rxin** [comment]
also this should probably go into sql?

File: project/SparkBuild.scala (hunk 2 above)
Comment (JoshRosen) [inline project/SparkBuild.scala]: Yes. The same goes for `sketch` and `mllibLocal`. Can you file a blocker JIRA for 2.0.1 / 2.1 to make sure that we audit this post-release?

File: dev/sparktestsupport/modules.py [hunk 4]
```diff
@@ -446,6 +446,17 @@ def __hash__(self):
     ]
 )
 
+compatibility = Module(
+    name="compatibility",
+    dependencies=[],
```
Comment (JoshRosen) [inline dev/sparktestsupport/modules.py]: We might want to pre-emptively add `hive` to this list.

**Comment by srowen** [comment]
Broadly, why is this a new and generic module? it seems quite Hive specific. I get that other fairly different compatibility shims could go here later, but those would properly be different modules right?

**Comment by yhuai** [comment]
@srowen I am moving this into sql.

**Comment by yhuai** [comment]
@JoshRosen https://issues.apache.org/jira/browse/SPARK-14818 is the jira for updating mima exclusions.

**Comment by yhuai** [comment]
@srowen @rxin @JoshRosen  How's this version?

File: sql/hive-1-x-compatibility/pom.xml [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+                        </javacArgs>
+                    </configuration>
+                </plugin>
+                <plugin>
+                    <groupId>org.apache.maven.plugins</groupId>
+                    <artifactId>maven-compiler-plugin</artifactId>
+                    <configuration>
+                        <compilerArgs combine.children="append">
+                            <!-- This option is needed to suppress warnings from sun.misc.Unsafe usage -->
+                            <arg>-XDignore.symbol.file</arg>
```
Comment (holdenk) [inline sql/hive-1-x-compatibility/pom.xml]: Supressing warnings for sun.misc.unsafe sounds good, but do we need it in the hive compatibility shims project (this does suppress other warnings that we might want to see if we aren't planning on using sun.misc.unsafe here)?

**Comment by JoshRosen** [comment]
LGTM. We can always change the name later.

**Pull request closed without merging**